qiskit/transpiler/passes/optimization/template_matching/template_substitution.py

```diff
@@ -500,39 +502,51 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
             template_params += template_dag_dep.get_node(t_idx).op.params
 
         # Create the fake binding dict and check
-        equations, symbols, sol, fake_bind = [], set(), {}, {}
-        for t_idx, params in enumerate(template_params):
-            if isinstance(params, ParameterExpression):
-                equations.append(sym.Eq(parse_expr(str(params)), circuit_params[t_idx]))
-                for param in params.parameters:
-                    symbols.add(param)
-
-        if not symbols:
+        equations, circ_dict, temp_symbols, sol, fake_bind = [], {}, {}, {}, {}
+        for t_idx, template_params in enumerate(template_params):
+            if isinstance(template_params, ParameterExpression):
+                if isinstance(circuit_params[t_idx], ParameterExpression):
+                    circ_param_sym = circuit_params[t_idx].get_sympy_expr()
+                else:
+                    circ_param_sym = parse_expr(str(circuit_params[t_idx]))
+                equations.append(sym.Eq(template_params.get_sympy_expr(), circ_param_sym))
+
+                for param in template_params.parameters:
+                    if not HAS_SYMENGINE:
+                        temp_symbols[param] = sym.Symbol(str(param))
+                    else:
+                        temp_symbols[param] = symengine.Symbol(str(param))
+
+                if isinstance(circuit_params[t_idx], ParameterExpression):
+                    for param in circuit_params[t_idx].parameters:
+                        if not HAS_SYMENGINE:
+                            circ_dict[param] = sym.Symbol(str(param))
+                        else:
+                            circ_dict[param] = symengine.Symbol(str(param))
+
+        if not temp_symbols:
             return template_dag_dep
 
         # Check compatibility by solving the resulting equation
-        sym_sol = sym.solve(equations)
+        sym_sol = sym.solve(equations, set(temp_symbols.values()))
         for key in sym_sol:
             try:
-                sol[str(key)] = float(sym_sol[key])
+                sol[str(key)] = ParameterExpression(circ_dict, sym_sol[key])
             except TypeError:
                 return None
 
         if not sol:
             return None
 
-        for param in symbols:
-            fake_bind[param] = sol[str(param)]
+        for key in temp_symbols:
+            fake_bind[key] = sol[str(key)]
 
         for node in template_dag_dep.get_nodes():
             bound_params = []
-
             for param in node.op.params:
                 if isinstance(param, ParameterExpression):
-                    try:
-                        bound_params.append(float(param.bind(fake_bind)))
-                    except KeyError:
-                        return None
+                    for key in fake_bind:
+                        bound_params.append(param.assign(key, fake_bind[key]))
```

Comment:
`for key, value in fake_bind.items():`, but also, isn't this going to add a whole lot of partially bound parameters to the `bound_params` list?  If `fake_bind` has more than one item in it, this is going to be called a whole lot of times, and there'll be more items in `bound_params` than there originally were in `node.op.params`.

Reply:
I think the issue here is that we cannot pass `bound_params` to `.bind()` because of the requirement the bound values are numeric, whereas we cannot pass `bound_params` directly to `.assign()` because it takes `parameter` and `value` separately as arguments. Any obvious solution seems to require modification of how `.bind()` or `.assign()` work.

Reply:
I don't think this comment matches up with my issue (also the point about iterating through a dictionary with `.items()` still stands).  I think this code is buggy, and it's just luckily not triggered.

The problem is that if `fake_bind` has `n` elements in it, the length of `bound_params` would be `n * len(params)`, which is incorrect.  Logically, this code is trying take `param`, and bind all instances of `Parameter` within it to a corresponding value in `fake_bind`, and return a _single_ value, but instead it adds lots of values to `bound_params`.  This only works because the templates you're considering only have a single `Parameter` in them.  If we had a template that had two parameters in it, this would be broken.  It's quite possible it was broken before as well, but this breaks it in a new way.

Please can you add a test case that involves a template (arbitrary - it can just be invented for the test) that has two parameters, and test that it binds correctly to input circuits with either numeric values or parameter values?  For example, here's a (nonsensical) template match that fails with a `KeyError`, even though it should bind just fine:
```python
In [5]: from qiskit.circuit import Parameter, QuantumCircuit
   ...: from qiskit.transpiler import PassManager
   ...: from qiskit.transpiler.passes import TemplateOptimization, RZXCalibrationBuilder
   ...:
   ...: a = Parameter('a')
   ...: b = Parameter('b')
   ...:
   ...: qc = QuantumCircuit(2)
   ...: qc.p(0.1, 0)
   ...: qc.p(-0.2, 1)
   ...:
   ...: template = QuantumCircuit(2)
   ...: template.p(a, 0)
   ...: template.p(b, 1)
   ...: template.cx(0, 1)
   ...:
   ...: pass_ = TemplateOptimization(
   ...:     user_cost_dict={"cx": 0, "p": 0},
   ...:     template_list=[template],
   ...: )
   ...: PassManager(pass_).run(qc).draw()
```

There are two issues at play: one is the issue with the length of `bound_params`, which is new in this PR, and the other is that the calls to `ParameterExpression.assign` during the template match should assign all parameters in that expression in one go, and _only_ those parameters (this issue I think was pre-existing).

Reply:
If it _really_ isn't possible to have it work for multiple parameters in the template circuit, then this should at least _enforce_ that the template circuit only has a single unbound parameter in it, and the pass should loudly fail if it's given a template with more than one parameter.